Lecture 5

cmakrr opened the pull request.

Commit:
Task 5.1: created checkout page

Commit:
Task 5.2: added order overview page

Commit:
Task 5.3: added dos protection

Commit:
Added tests for lecture 5

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/dao/order/OrderDao.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Optional;
+
+public interface OrderDao extends Dao<Order> {
+
```
placeOrder?

cmak33 replied:
what if my placeOrder will only call save method?

esSoftPalinaTsi replied:
fine option

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/order/Order.java:
```diff
@@ -0,0 +1,37 @@
+@Setter
+public class Order extends Cart implements Entity {
+
+    private Long id;
+    private String secureId;
```
why  do we need 2 id-s

cmak33 replied:
Overview page displays private information (phone number, adress, surname ...), so when we use long in overview url it is pretty easy to guess some valid overview id and check private info, with String id it is less likely, because it is much harder to guess

esSoftPalinaTsi replied:
Not sure abiut "easy to guess the valid id"
But anyway then we can use one secruId everywhere, no need to ovecomplicate model with many unique fileds

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/Product.java:
```diff
@@ -12,7 +13,7 @@
 @Setter
 @EqualsAndHashCode
-public class Product {
+public class Product implements Entity {
```
why implements
and where is Serializable

cmak33 replied:
I decided to make Entity an interface , because everything we need from entity in our dao (basic interface) is to get entity id, and it allows us to make any class an entity , even if it extends another class (like in case with order that extends cart)

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/validator/parameterValidator/PhoneParameterValidator.java:
```diff
@@ -0,0 +1,44 @@
+                    return true;
+                }
+            } catch (NumberParseException ignored) {
+            }
```
we are ignoring exception?

cmak33 replied:
We will duplicate code (at least one line that is responsible for putting error into map) if we try to handle it

esSoftPalinaTsi replied:
Between catching and doing nothing and duplicating one line the second option is less harmful
If we have such empty catches, then it means that 1. we shouldn't catch or 2. we should do smth inside the catch

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/validator/parameterValidator/PhoneParameterValidator.java:
```diff
@@ -0,0 +1,44 @@
+        String PHONE_PATTERN = "[+]?[\\d\\s]+";
+        if (value.matches(PHONE_PATTERN)) {
+            try {
+                Phonenumber.PhoneNumber number = phoneNumberUtil.parse(value, null);
```
why null

cmak33 replied:
The second parameter is phone number region, when null is passed it parses number for any region/country

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/validator/localDateValidator/DeliveryDateValidator.java:
```diff
@@ -0,0 +1,32 @@
+package com.es.phoneshop.model.validator.localDateValidator;
```
all validators should be outside the model package

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/cart/ProductAndQuantity.java:
```diff
@@ -2,5 +2,5 @@
 import com.es.phoneshop.model.product.Product;
 
-public record CartProduct(Product product, int quantity) {
+public record ProductAndQuantity(Product product, int quantity) {
```
why
previous name was better

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/order/CustomOrderService.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public void placeOrder(Order order) {
+        order.generateSecureId();
```
service does nothing with model attributes

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/filters/DosProtectionFilter.java:
```diff
@@ -0,0 +1,36 @@
+        if (dosProtectionService.isUserAllowed(ip)) {
+            filterChain.doFilter(servletRequest, servletResponse);
+        } else {
+            int TOO_MANY_REQUESTS_STATUS = 429;
```
we should have reserved constant for this status

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/servlets/CartServlet.java:
```diff
@@ -35,7 +35,7 @@ public void init(ServletConfig config) throws ServletException {
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
         AttributesHolder attributesHolder = new HttpSessionAttributesHolder(request.getSession());
-        request.setAttribute("products", cartService.getCartProducts(attributesHolder));
+        request.setAttribute("products", cartService.getCartProductsFromAttributesHolder(attributesHolder));
```
getCartProducts

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/pages/checkout.jsp:
```diff
@@ -0,0 +1,105 @@
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+<%@ taglib prefix="tags" tagdir="/WEB-INF/tags" %>
+
+<jsp:useBean id="isEmpty" type="java.lang.Boolean" scope="request"/>
```
it's not java bean

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/servlets/CheckoutPageServlet.java:
```diff
@@ -0,0 +1,136 @@
+            request.setAttribute("paymentMethods", PaymentMethod.getMethodNames());
+            request.setAttribute("isEmpty", false);
+        } else {
+            request.setAttribute("isEmpty", true);
```
we stored cart in session, so the fact of emptiness of the cartItems can be checked directly in jsp

cmak33 replied:
We can check it in jsp, but we have to check cart emptiness in servlet anyway (to avoid creation of unnecessary order object, and no to put unnecessary attributes to request), so we will check cart emptiness twice

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/dao/CustomOrderDaoTest.java:
```diff
@@ -0,0 +1,51 @@
+            add(order);
+        }});
+    }
+
```
tests for save() and getEntity()?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/validator/parameterValidator/NotBlankParameterValidator.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.Map;
+
+public class NotBlankParameterValidator implements ParameterValidator {
```
not good enough to create a name based on you implementation details
now you're checking only isBlank(), but if in future you need a check for maxLength, then the validator will be renamed too
some general name would be better

Commit:
Fixed issues:
- added placeOrder method to OrderDao
- moved validator package from model
- renamed ProductAndQuantity to CartProduct
- added tests for save and getEntity methods of OrderDao
- renamed NotBlankParameterValidator to DefaultParameterValidator

Commit:
Fixed issues:
- added generic id parameter to Entity
- refactor validate method for PhoneParameterValidator not to ignore exception
- removed unused long id from Order

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/entity/Entity.java:
```diff
@@ -1,6 +1,8 @@
+import java.io.Serializable;
 
-    Long getId();
+public interface Entity<T> extends Serializable {
```
extends?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/order/Order.java:
```diff
@@ -13,10 +13,9 @@
 
-    private Long id;
-    private String secureId;
+    private String id;
```
still don't get why we can't leave long as id
you generate the value via UUID, it's a random value, no matter if it's long or string
in all previous pojo-s your id was long

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/dao/CustomOrderDaoTest.java:
```diff
@@ -6,21 +6,23 @@
 
     @Mock
     private Order order;
+    @Spy
```
what is the reason of spy here?

cmak33 replied:
in placeOrder test i verify that save method of dao was called

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/dao/genericDao/GenericDao.java:
```diff
@@ -0,0 +1,72 @@
+
+@Getter(AccessLevel.PROTECTED)
+@Setter
+public class GenericDao<A, T extends Entity<A>> implements Dao<A, T> {
```
should be abstract by it's purpose

Commit:
Fixed issues:
- made GenericDao abstract
- made Order id long
- made Entity not to extend Serializable

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/exception/notFoundException/OrderNotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+public class OrderNotFoundException extends EntityNotFoundException {
+
+    public OrderNotFoundException(Long id) {
+        super(id,"Order");
```
formatting

esSoftPalinaTsi commented:
Message in minicart.jsp is too long
In chrome it happens that it is displaying under the Phoneshop

esSoftPalinaTsi commented:
Message in minicart.jsp is too long
In chrome it happens that it is displaying under the Phoneshop

esSoftPalinaTsi commented:
Add stock decrease after the placing order

Commit:
Fixed issues:
- changed minicart page text
- made product stock decrease after placing order
- fixed formatting in OrderNotFoundException
